src/ast/irbuilderbpf.cpp

```diff
@@ -830,6 +830,101 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
   return CreateHelperCall(fn, gettime_func_type, {}, "get_ns", &loc);
 }
 
+Value *IRBuilderBPF::CreateIntegerArrayCmp(Value *ctx,
+                                           Value *val1,
+                                           Value *val2,
+                                           const SizedType &val1_type,
+                                           const SizedType &val2_type,
+                                           const bool inverse,
+                                           const location &loc)
+{
+  /*
+   // This function compares each character of the two arrays.
+   // It returns true if all are equal and false if any are different
+   // cmp([]char val1, []char val2)
+   {
+      for (size_t i = 0; i < n; i++)
+      {
+        if (val1[i] != val2[i])
+        {
+          return false;
+        }
+      }
+      return true;
+   }
+*/
+  auto elem_type = *val1_type.GetElementTy();
+  const uint64_t n = val1_type.GetNumElements();
+
+  Value *ll, *rr, *cmp;
+  AllocaInst *l = CreateAllocaBPF(elem_type, "ll");
+  AllocaInst *r = CreateAllocaBPF(elem_type, "rr");
+  Function *parent = GetInsertBlock()->getParent();
+  AllocaInst *store = CreateAllocaBPF(getInt1Ty(), "arraycmp.result");
+  BasicBlock *arr_ne = BasicBlock::Create(module_.getContext(),
+                                          "arraycmp.false",
+                                          parent);
+  BasicBlock *done = BasicBlock::Create(module_.getContext(),
+                                        "arraycmp.done",
+                                        parent);
+
+  CreateStore(getInt1(!inverse), store);
+  for (size_t i = 0; i < n; i++)
+  {
+    BasicBlock *arr_eq = BasicBlock::Create(module_.getContext(),
```

Comment:
Similarly to the strstr discussion in #2393, I think this implementation would be better if we don't unroll the loop.

Reply:
Seems reasonable to me to gate this helper based on bounded loop presence.

Reply:
> based on bounded loop support

That would be.. [5.3, the patch is from Jun 2019](https://lwn.net/Articles/794934/). Runtime checks for this are annoying but possible.

Reply:
You mean a feature probe like this? https://github.com/iovisor/bpftrace/blob/86b7843b0e56c96dc91c9fa0354ffe1507245ad8/src/bpffeature.cpp#L191-L207